proposals/0000-local-do.rst

```diff
@@ -52,6 +52,296 @@ This proposal creates a new language extension ``-XQualifiedDo``.
 
 When ``-XQualifiedDo`` is activated, the syntax of the ``do`` notation is changed to
 
+::
+
+  <lexp> ⟶ [modid.]do
+
+where ``modid`` stands for some module name.
+
+The additional module name is called the qualifier of the do-expression.
+
+The semantics of ``do`` notation statements is given as follows (using
+``-XLambdaCase`` notation):
+
+* The ``x <- u`` statement uses ``(modid.>>=)``
+
+  ::
+
+    M.do { x <- u; stmts }  =  u M.>>= \x -> M.do { stmts }
+
+* The ``u`` statement uses ``modid.(>>)``
+
+  ::
+
+    M.do { u; stmts }  =  u M.>> M.do { stmts }
+
+* The a ``pat <- u`` statement uses ``M.fail`` for the failing case,
+  if such a case is needed
+
+  ::
+
+    M.do { pat <- u; stmts }  =  u M.>>= \case
+      { pat -> M.do { stmts }
+      ; _ -> M.fail "…"
+      }
+
+  If the pattern cannot fail, then we don't need to use ``M.fail``.
+
+  ::
+
+    M.do { pat <- u; stmts }  =  u M.>>= \case pat -> M.do { stmts }
+
+* ``-XApplicativeDo`` uses ``(M.<$>)``, ``(M.<*>)`` and ``M.join`` (this
+  assumes that the applicative-do grouping has been performed)
+
+  ::
+
+    M.do { (x1 <- u1 | … | xn <- un); return e }  =
+      (\x1 … xn -> e) M.<$> u1 M.<*> … M.<*> un
+
+    M.do { (x1 <- u1 | … | xn <- un); stmts }  =
+      M.join (\x1 … xn -> M.do { stmts }) M.<$> u1 M.<*> … M.<*> un
+
+
+  Note that ``M.join`` is only needed if the final expression is
+  not identifiably a ``return``.
+
+  When the applicative statements contain nested statements (see the
+  `wiki page
+  <https://gitlab.haskell.org/ghc/ghc/wikis/applicative-do>`_ for a
+  complete description of applicative-do statements), we also need a
+  ``M.return``. *e.g.*
+
+  ::
+
+    M.do { ({stmt1; …; stmtn} {x1; …; xn} | y <- u) ; return e }  =
+      (\(x1,…,xn) y -> e) <$> (M.do { stmt1; …; stmtn; M.return (x1, …, xn)}) <*> u
+
+*  With ``-XRecursiveDo``, ``rec`` blocks use ``M.mfix`` and ``M.return``:
+
+   ::
+
+     M.do { rec { x1 <- u1; … ; xn <- un }; stmts }  =
+       M.do
+       { (x1, …, xn) <- M.mfix (\~(x1, …, xn) -> M.do { x1 <- u1; …; xn <- un; M.return (x1, …, xn)})
+       ; stmts
+       }
+
+If a name ``M.op`` is required by the desugaring process (and only if it's required!) but the name is not in scope, an error message is produced:
```

Comment:
```suggestion
If a name ``M.op`` is required by the desugaring process (and only if it's required!) but the name is not in scope, an error message like the following is produced:
```
(Of course GHC is free to produce a better error message, e.g. tell the user if the thing exists as `M.op`, but has not been imported.)